// @jupyterlab/latex/lib/error.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
import { Widget } from '@phosphor/widgets';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
const LATEX_ERROR_PANEL = 'jp-LatexErrorPanel';
const LATEX_ERROR_CONTAINER = 'jp-LatexErrorContainer';
/**
 * A widget which hosts the error logs from LaTeX
 * when document compilation fails.
 */
export class ErrorPanel extends Widget {
    /**
     * Construct the error panel.
     */
    constructor() {
        super();
        this.addClass(LATEX_ERROR_PANEL);
    }
    set text(value) {
        ReactDOM.render(React.createElement(LatexError, { text: value }), this.node, () => {
            this.update();
        });
    }
    /**
     * Handle an update request.
     */
    onUpdateRequest(msg) {
        const el = this.node.children[0];
        el.scrollTop = el.scrollHeight;
    }
    /**
     * Handle a close request.
     */
    onCloseRequest(msg) {
        this.dispose();
    }
}
export class LatexError extends React.Component {
    render() {
        return (React.createElement("pre", { className: LATEX_ERROR_CONTAINER },
            React.createElement("code", null, this.props.text)));
    }
}
